engine/crates/federated-graph/src/federated_graph/v2.rs

```diff
@@ -0,0 +1,517 @@
+pub use super::v1::{
+    Definition, EnumId, FieldId, FieldProvides, FieldRequires, FieldSet, FieldSetItem, FieldType, FieldTypeId,
+    InputObjectId, InterfaceField, InterfaceId, Key, ListWrapper, ObjectField, ObjectId, Override, OverrideSource,
+    RootOperationTypes, ScalarId, StringId, Subgraph, SubgraphId, UnionId,
+};
+
+/// A composed federated graph.
+///
+/// ## API contract
+///
+/// Guarantees:
+///
+/// - All the identifiers are correct.
+///
+/// Does not guarantee:
+///
+/// - The ordering of items inside each `Vec`.
+#[derive(serde::Serialize, serde::Deserialize, Clone)]
+pub struct FederatedGraphV2 {
+    pub subgraphs: Vec<Subgraph>,
+
+    pub root_operation_types: RootOperationTypes,
+    pub objects: Vec<Object>,
+    pub object_fields: Vec<ObjectField>,
+
+    pub interfaces: Vec<Interface>,
+    pub interface_fields: Vec<InterfaceField>,
+
+    pub fields: Vec<Field>,
+
+    pub enums: Vec<Enum>,
+    pub unions: Vec<Union>,
+    pub scalars: Vec<Scalar>,
+    pub input_objects: Vec<InputObject>,
+
+    /// All [input value definitions](http://spec.graphql.org/October2021/#InputValueDefinition) in the federated graph. Concretely, these are arguments of output fields, and input object fields.
+    pub input_value_definitions: Vec<InputValueDefinition>,
+
+    /// All the strings in the federated graph, deduplicated.
+    pub strings: Vec<String>,
+
+    /// All the field types in the federated graph, deduplicated.
+    pub field_types: Vec<FieldType>,
+
+    /// All composed directive instances (not definitions) in a federated graph.
+    pub directives: Vec<Directive>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct InputValueDefinition {
+    pub name: StringId,
+    pub field_type_id: FieldTypeId,
```

Comment:
Up to you, I've named it `TypeId` in the `engine-v2-schema` because it mixes Input & Output types. An input value can also be an argument, so field_type  feels a bit weird imho

Reply:
Yes, it was to avoid the confusion between an identifier for a top level type and an identifier for the type of a field (which this is). Taking from the spec, maybe `TypeRefId` would be even better, but let's go with TypeId since it's shorter and the other interpretation of typeid already exists and it's called `Definition`.